source/extensions/filters/network/thrift_proxy/filters/payload_to_metadata/payload_to_metadata_filter.h

```diff
@@ -135,15 +135,16 @@ class TrieMatchHandler : public DecoderCallbacks,
   }
 
   FilterStatus handleContainerEnd() {
+    ASSERT(steps_ > 0, "unmatched container end");
     steps_--;
     return FilterStatus::Continue;
   }
 
   MetadataHandler& parent_;
   TrieSharedPtr node_;
   bool complete_{false};
-  absl::optional<int16_t> last_field_id_;
-  uint16_t steps_{0};
+  std::vector<int16_t> field_ids_;
+  int16_t steps_{0};
```

Comment:
Is there a reason this can never be > 32767? Wonder if we should check for overflow or make this a bigger value?

Reply:
good question. `steps_` is to track how many layers we've entered which is not fit the config. i.e., we'll hit the large number only if we have struct like
```
struct A1 {
  A2 a2;
}

struct A2 {
 A3 a3;
}

...

struct A32767 {
 A32768 a32768;
}

```

and thrift schema doesn't prevent this

it might be possible to hit this if we have a recursive field most likely so I'll propose to 
(a) report [on_missing](https://github.com/envoyproxy/envoy/blob/main/api/envoy/extensions/filters/network/thrift_proxy/filters/payload_to_metadata/v3/payload_to_metadata.proto#L84) in release
(b) assert steps_ < 30000 
(c) note this in doc

in next PR